### DISCUSSION TITLE
Are the Queries run in coroutines? Or, do I need to call them under a coroutine?

### DISCUSSION THREAD

**Discussion by meta-forte** [vincentlauvlwj/Ktorm#537] [Q&A] [answered]
Does ktorm run SQL statements & queries by starting a coroutine? or, do I need manage execution of queries under coroutines myself?

thanks

**Comment by JasonMing** [comment, marked as answer]
AFAIK, ktorm is running in blocking mode(see: org.ktorm.database.Database#executeExpression), the actually query will block the caller's thread.

Which means, ktorm queries should be dispatched to a standalone blocking dispatcher. Especially the operations depend on the transaction, because `JdbcTransactionmanager` depends on `ThreadLocal`.

Here is what I do:

```kotlin
private val dbDispatcher = newFixedThreadPoolContext(30, "ktorm-db")

suspend fun <R> db(block: suspend Database.() -> R): R =
    withContext(dbDispatcher) {
        block(db)
    }
```

So you can simply call `db { ... }` in any suspend function without worrying about blocking the non-blocking coroutine dispatchers.
